design/one-pager-generic-datasource.md

```diff
@@ -0,0 +1,189 @@
+# Patch from Generic Datasources
+
+* Owner: Maximilian Blatt (maximilian.blatt@accenture.com, @mistermx)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane currently supports patching from composite resources to composed
+resources and vice versa. It is also possible to copy values between two
+composed resources inside one compositiion by patching the value from one
+composed back to the composite and then to the second composed resource.
+
+However, Crossplane currently does not provide a way to patch from
+environment-dependent data sources. 
+Compositions behave the same regardless in which environment they are deployed.
+
+This becomes a big issue once you want to deploy the same composition in
+multiple environments. For example deploying AWS resources in two different VPCs
+requires different subnet IDs, VPC Ids, OIDC issuers etc.
+
+## Current workarounds
+
+### Build one composition per environment
+
+The simplest solution would be to build one composition per environment.
+While this indeed works, it becomes increasingly complex with the number of
+environments.
+
+### Use generic provider referencers
+
+While compositions itself are not environment-dependent, some providers serve
+some MRs that contain generic referencers. 
+
+* `provider-helm`: using [`spec.forProvider.valuesFrom.secretKeyRef`](https://doc.crds.dev/github.com/crossplane-contrib/provider-helm/helm.crossplane.io/Release/v1beta1@v0.10.0#spec-forProvider-valuesFrom-secretKeyRef)
+* `provider-kubernetes`: using [`spec.forProvider.references[].patchesFrom`](https://doc.crds.dev/github.com/crossplane-contrib/provider-kubernetes/kubernetes.crossplane.io/Object/v1alpha1@v0.3.0#spec-references-patchesFrom)
+
+However, none of these methods are Crossplane native and are not available to
+every provider. Furthermore, they would require an additional MR (here `Object`
+or `Release`) to deploy the composed resource increasing the complexity of the
+composition.
+
+## Goals
+
+The goal of this document is to define a proposal for extracting environment
+specific data considering the following requirements:
+
+1. Allow the creation of generic compositions that are able to load environment 
+specific parameters during compose time.
+2. Allow users to define these values on cluster level.
+3. Prevent users from escalating their rights on the cluster and read out fields
+from objects they shouldn't have access to.
+4. (Bonus) Allow package writers to include default values in configuration
+packages.
+
+## Proposed Solution
+
+Introduce a new Crossplane CRD `ConfigSet` that can store arbitrary data similar
+to K8s `ConfigMap` but supports complex data using the [`JSON` type](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1#JSON) that can be anything that is valid JSON:
```

Comment:
@liggitt just reminded me that the whole point of Secrets and ConfigMaps was to expose their data as files and/or env vars. Given we don't have that constraint here supporting complex data makes sense to me.

Reply:
> Can you outline some use cases for storing complex data, rather than a map[string][]byte?

For example, in our case, we want to patch a list of subnet IDs directly into a resource without picking them one by one. Additionally, have a hierarchical structure allows you to order stuff that belongs together, i.e `aws.subnetsIds`  and `azure.subnetIds` instead of `aws_subnet_1` ,`azure_subnet_2` etc.

> it does have a somewhat more predictable, self-documenting schema

My idea behind `ConfigSets`s it is to _not_ restrict the user with a schema but rather let him chose the data structure on his own.

Reply:
Sounds good - let's proceed with unstructured complex data.